README.md
# Vertical Tabs Plus package
Makes tabs vertical and adds some extra features. Fork from [tiger4th's vertical-tabs package](https://atom.io/packages/vertical-tabs).

## Features

### Switch tabs by scrolling
Move cursor over tabs and switch between them using scroll wheel. Inspired by the same named option of the Vivaldi browser.

### Tabs placement: left, right, embedded (above project tree)
Tabs can be placed on either left or right side of the main pane. It's also possible to shelter them above the project tree not to make it waste much space. Inspired by Adobe Brackets interface.

### Customizable tab width and height
Vertical tabs are not hosintal tabs so it makes sense to adjust their sizes manually. When Atom 1.17 has been released, the default theme ("One") got tiny tabs. Some people liked it, some were looking for the way to bring spacious tabs back and made a number of custom styles. One more reason why this option is more important than it seems :)  
Tab width is also customizable or, more accurately, its limits are. Atom adjusts tab width itself all the time, this package just set min and max values.

## Why forked?
Originally I had a bunch of custom styles that were fixing weird things in the original package like space under each tab and wrong styles that are used to be actual for vertical tabs only (extra border on each left tab side except the first one, double borders at the bottom of the tab container, etc.). Then I put some own code into the init script file to switch between tabs with scroll wheel (blame Vivaldi and QTTabBar, they made me love it). Later I completely rewrote styles, added custimizable tab height (via LESS variable) and vertical tabs stopped look weird when themes like Material are used. And then I found a way to move tabs over the project tree view what required even more JS code. So I moved everything to a separate package.

There are too many changes for a single pull request but it's still based on the tiger4th's package so I decided to fork it. Well, people fork projects for purposes like this so why not?

The difference from the original package:
* Works well with "Material" theme, other themes shouldn't break it either (at the time of development is tested with "[One](https://atom.io/themes/one-light-ui)", "[Nord](https://atom.io/themes/nord-atom-ui)", and "[Material](https://atom.io/themes/atom-material-ui)");
* Replaced fixed tabs width with `min` and `max` values;
* Can be placed over the project tree for sake of saving space;
* Customizable tab properties (width, height, and intends);
* Extra features.

## Why 2.0.0, not 0.1.0?
Because it's a forked project and versions up to 1.0.5 are already taken by the original package:
```
npm ERR! Command failed: git -c core.longpaths=true tag v0.1.0 -am Prepare 0.1.0 release
npm ERR! fatal: tag 'v0.1.0' already exists
```

## Known problems:
* When dragging tabs, the tab placeholder is displayed _above_ a tab if cursor is over left side of tab container. As it's displayed _below_ a tab for the right tab container side. It's natural behavior for horizontal tabs but looks weird for vertical ones. Sadly it's a not fixable problem.

## Screenshots
Tabs on the left side:
![](https://github.com/OddMorning/vertical-tabs-plus/raw/master/screenshots/left.png)

Tabs on the right side:
![](https://github.com/OddMorning/vertical-tabs-plus/raw/master/screenshots/right.png)

Tabs above project folders:
![](https://github.com/OddMorning/vertical-tabs-plus/raw/master/screenshots/embedded.png)

CHANGELOG.md
# Changelog

## 2.4.0 (04 Apr 2018)
* **Added:** Click on file in project tree won't shift project tree down immediately when `Embedded` tab placement is chosen (similar to browser's behavior when tabs get closed). This allows to click a file twice to pin it (and it simply looks nice).

#### 2.3.5 (12 Mar 2018)
* **Fixed:** The package didn't process error properly in case of failed attaching tabs to project tree
* Renamed "Over project tree" tabs placement into "embedded" placement
* Minor internal changes

#### 2.3.4 (24 Nov 2017)
* **Fixed:** Max tab container height didn't work properly with multiple panes.

#### 2.3.3 (22 Nov 2017)
* **Improved:** `Fix Tooltips Placement` doesn't show tooltips in wrong direction anymore. This option looks more or less stable now;
* **Improved:** Better changelog design.

#### 2.3.2 (21 Nov 2017)
* **Improved:** The `Over project tree` tab position is stable now, tabs like Settings or Diagnostics can be moved and sorted safely as they won't try to dock anymore;
* **Fixed:** The close button became narrow when tab has long title.

#### 2.3.1 (19 Nov 2017)
* **Improved:** The `Tab scroll` option is stable and works independently on focused pane;
* **Fixed:** Multiple panes are supported now. They were supported in the original package so technically it's just a bug fix;
* **Fixed:** Some tabs didn't show tooltips properly with `Fix Tooltips Placement` option enabled.

## 2.3.0 (14 Nov 2017)
* **Improved:** Redesigned settings menu, options are grouped now;
* **Fixed** some trifles, internal changes;
* **Added** an experimental ability to fix tooltip placement. By default it's displayed below tabs that doesn't look right for vertical tabs and this option moves tooltips to the right of the tab instead of below it (or to the left when the tabs are on the right side). This option is experimental and not really stable because there's no way to inject into the "tabs" package and change behaivor of tooltips so it tries to intercept tooltips before they're displayed instead. This is the best way that I could think of. Tooltips still appear below tabs if you move cursor too quick before they were displayed for the first time.  
The idea [is suggested by kreba](https://github.com/tiger4th/vertical-tabs/issues/19).

## 2.2.0 (10 Nov 2017)
* **Added** a new option to control how much space tabs can take when placed over project tree (60% by default);
* **Fixed** package crash when a new window opened with the `Over project tree` tabs placement chosen (tabs may blink while waiting for the project tree pane since it's hidden in new windows by default);
* **Fixed** problems with scrolling when tabs don't fit the screen and switching tabs by scrolling option is enabled;

#### 2.1.1 (09 Nov 2017)
* **Fixed:** Scrollbar didn't show tabs when they don't fit the screen.

## 2.1.0 (09 Nov 2017)
* **Added** an ability to customize tab height and intends manually;
* **Fixed** bugs with shrinking tabs down to 0px with the `Over project tree` tabs placement chosen.

## 2.0.0 (08 Nov 2017)
* Completely rewritten styles and code, still based on the original tiger4th's package;
* Corrected small interface flaws like weird space between tabs, 1px shift when dragging tabs, missing left border for a first tab, double borders in other places and so on;
* Added an ability to switch tabs by scrolling (inspired by the same named option of the Vivaldi browser);
* Tabs can be placed over the projects folders (a.k.a. tree view). It's still beta so there can be bugs;
* Works best with the default theme (One Light/Dark) but should work with other themes too. Is tested with [Nord](https://atom.io/themes/nord-atom-ui) and [Material](https://atom.io/themes/atom-material-ui) themes;
* It's the first release of this forked package so there still can be bugs and flaws.

#### 0.1.0 - 1.0.5
* Original package development by tiger4th, the changelog is [available on GitHub](https://github.com/tiger4th/vertical-tabs/blob/v1.0.5/CHANGELOG.md)

lib/config.js
'use babel';

export default {
  general: {
    type:       'object',
    order:       1,
    properties: {
      tabsPlacement: {
        title:   'Tabs Placement',
        type:    'string',
        default: 'right',
        enum:    [
          {value: 'left', description: 'Left'},
          {value: 'right', description: 'Right'},
          {value: 'tree-view', description: 'Embedded'},
        ],
        order:   1,
      },
      scrollTabs: {
        title:       'Scroll Tabs',
        type:        'boolean',
        default:     true,
        description: 'Change active tab by mouse wheel over tabs',
        order:       2,
      },
      normalizeTooltips: {
        title:       'Fix Tooltips Placement',
        type:        'boolean',
        default:     true,
        description: 'Moves tooltips to the left or right side (depends on tab placement). Since it\'s impossible to change tooltip position the legal way, this option uses some workarounds',
        order:       4,
      },
    },
  },
  tab: {
    type:       'object',
    order:       2,
    properties: {
      height: {
        title:       'Tab Height',
        type:        'string',
        default:     '2.9em',
        enum:    [
          {value: '2em', description: 'Dense (2em)'},
          {value: '2.5em', description: 'Compact (2.5em)'},
          {value: '2.9em', description: 'Medium (2.9em)'},
          {value: '3.25em', description: 'Spacious (3.25em)'},
          {value: 'custom', description: 'Custom'},
        ],
        description: 'Atom before version 1.17 had `Spacious` tabs and since 1.17 they became `Compact`',
        order:       1,
      },
      customHeight: {
        title:       'Custom Tab Height',
        type:        'string',
        default:     '2.5',
        description: 'Any valid CSS value is allowed. Number without any unit is considered as `em`. You have to chose `Custom` tab height to make custom options work',
        order:       2,
      },
      customIndent: {
        title:       'Custom Tab Indent Coefficient',
        type:        'number',
        default:     '4',
        description: 'Tab left and right indents are calculated the following way: `tab-height / coefficient`. This options allows to set the `coefficient` value. For example, `2` means intends equal to half of tab height (`1.5em` indents for `3em` tab height). The default value for every stock tab height is equal to 4, for `Dense` is 2',
        order:       3,
      },
    },
  },
  tabContainer: {
    type:       'object',
    order:       3,
    properties: {
      minWidth: {
        title:       'Min Width',
        type:        'string',
        default:     '11',
        description: 'Any valid CSS value is allowed. Number without any unit is considered as `em` (`11` == `11em`). Use `initial` to disable limit (not recommended)',
        order:       1,
      },
      maxWidth: {
        title:       'Max Width',
        type:        'string',
        default:     '14',
        description: 'Note that these values are ignored when the `Over tree view` tabs placement is chosen',
        order:       2,
      },
      maxHeight: {
        title:       'Max Height',
        type:        'number',
        default:     60,
        description: "When placed above project tree, don't let tabs take more than % of dock height",
        order:       3,
      },
    },
  },
};